banzai_nres/rv.py

```diff
@@ -0,0 +1,77 @@
+from banzai.stages import Stage
+from banzai.frames import ObservationFrame
+import pkg_resources
+from astropy.io import fits
+import numpy as np
+from astropy.table import Table
+from astropy import constants
+
+# Speed of light in km/s
+c = constants.c.to('km / s').value
+
+
+def cross_correlate(velocities, wavelength, flux, flux_uncertainty, template_wavelength, template_flux):
+    """
+
+    :param velocities: in km / s
+    :param wavelength:
+    :param flux:
+    :param template_wavelength:
+    :param template_flux:
+    :return:
+
+    This is consistent with what we did in the satellite pipeline and
+    Zackay & Ofek, 2017, ApJ, 836, 187
+    """
+    x_cor = []
+    # for steps in 1 km/s from -2000 to +2000 km/s
+    # calculate the variance ahead of time
+    variance = flux_uncertainty * flux_uncertainty
+
+    for v in velocities:
+        doppler = 1.0 / (1.0 + v / c)
+        kernel = np.interp(doppler * wavelength, template_wavelength, template_flux)
+        correlation = kernel * flux
+        correlation /= variance
+
+        normalization = kernel * kernel / variance
+
+        # This gives us the signal to noise ratio in terms of N-sigma
+        snr = correlation.sum() / np.sqrt(normalization.sum())
+        x_cor.append(snr)
+
+    return np.array(x_cor)
+
+
+class RVCalculator(Stage):
+    TEMPLATE_FILENAME = pkg_resources.resource_filename('banzai_nres', 'data/g2v_template.fits')
+
+    def do_stage(self, image) -> ObservationFrame:
+        # Load in the template
+        template_hdu = fits.open(self.TEMPLATE_FILENAME)
+        ccfs = []
+        for i in image.fibers[image.fibers['fiber'] == image.science_fiber]['trace']:
```

Comment:
I don't think we should compute the CCF for (or at least not include in the mean RV computation) all of the orders, even for this first pass at RV computation. We already _know_ that some of the orders are going to be rendered useless by tellurics, at least without careful treatment that is beyond what we're trying to do at this stage. I think it would be better to just start with ~5-10 orders near the middle of the spectral range for testing purposes, and once we establish that that's working we can decide what spectral range we actually want to use for the RV computation.

Reply:
I agree. Tim Brown's recommendation was orders 25-50. In my experience, 25-35 (5-10 orders as Marshall suggests) is good as well.

I suggest that we code in a weights matrix to keep this general. We can assign weights of 0 to e.g. orders 0-24 and 51-67 . @cmccully @captain-exoplanet what do you think of this suggestion?

Reply:
That does sounds like a good general way to handle this.

Reply:
Is that order 25-50 the physical order? Or the order on the chip?

Reply:
On the weights matrix, I'm not really sure what that buys us. I'm really not seeing how this work in practice. The weights depend both on the template and the _velocity_ grid you are searching. If you loaded the weights earlier it would just split the logic into two places. The whole cross correlate function is only 17 lines of code (excluding the docstring) so I would think that you would just replace that if you wanted to test different weights. It would be a different question if we were using pixel sized steps in velocity space for example, but since we are subsampling the complexity goes up significantly to just store a matrix. Given all that, this feels a little premature to generalize to that level at least for this PR.

Reply:
25-50 on the physical order.